lib/api/include/irods/ticketAPI.hpp

```diff
@@ -0,0 +1,414 @@
+ *
+ */
+namespace irods::administration::ticket
+{
+    inline struct admin_tag
+    {
+    } admin;
+
+    enum class ticket_type
+    {
+        read,
+        write
+    };
+
+    enum class ticket_property
+    {
+        uses_count,
+        write_file,
+        write_byte
+    };
+
+    void execute_ticket_operation(RxComm& conn,
+                                  std::string_view command,
+                                  std::string_view ticket_identifier,
+                                  std::string_view command_modifier1,
+                                  std::string_view command_modifier2,
+                                  std::string_view command_modifier3,
+                                  std::string_view command_modifier4,
+                                  bool run_as_admin);
+
+    /**
+     * @brief Create a ticket object
+     *
+     * @param conn Communication object
+     * @param _type Specify type for ticket
+     * @param obj_path Path for the object
+     * @param ticket_name Name of the ticket (if provided)
+     * @return int Error code
+     */
+    void create_ticket(RxComm& conn, ticket_type _type, std::string_view obj_path, std::string_view ticket_name);
+    std::string create_ticket(RxComm& conn, ticket_type _type, std::string_view obj_path);
+
+    void create_ticket(admin_tag,
+                       RxComm& conn,
+                       ticket_type _type,
+                       std::string_view obj_path,
+                       std::string_view ticket_name);
+    std::string create_ticket(admin_tag, RxComm& conn, ticket_type _type, std::string_view obj_path);
+
+    /**
+     * @brief Delete a ticket
+     *
+     * @param conn Communication Object
+     * @param ticket_name Name of the ticket
+     * or
+     * @param ticket_ID ID for the ticket that needs to be deleted
+     * @return int Error Code
+     */
+    void delete_ticket(RxComm& conn, std::string_view ticket_name);
+    void delete_ticket(RxComm& conn, int ticket_ID);
+
+    void delete_ticket(admin_tag, RxComm& conn, std::string_view ticket_name);
+    void delete_ticket(admin_tag, RxComm& conn, int ticket_ID);
+
+    struct users_constraint
+    {
+        const std::string_view user_name;
+    };
+
+    struct groups_constraint
+    {
+        const std::string_view group_name;
+    };
+
+    struct hosts_constraint
+    {
+        const std::string_view host_name;
+    };
+
+    struct ticket_property_constraint
+    {
+        const ticket_property desired_property;
+        int value_of_property = -1;
+    };
+
+    template <typename T>
+    void addTicketConstraints(RxComm& conn, const std::string_view ticket_name, const T& constraints)
+    {
+        std::string_view command = "mod";
+        std::string_view command_modifier1 = "add";
+        std::string_view command_modifier2 = "";
+        std::string_view command_modifier3 = "";
+        std::string_view command_modifier4 = "";
```

Comment:
Why are these needed? They could be hard-coded directly into the call to `execute_ticket_operation`.